File: intrinsic_core/src/arm2d.cpp
#include "intrinsic_core/arm2d.hpp"

#include <algorithm>
#include <cmath>

namespace intrinsic_core
{

namespace
{
Eigen::Matrix<double, 9, 2> build_joint_actions()
{
  Eigen::Matrix<double, 9, 2> table;
  int row = 0;
  for (int a = -1; a <= 1; ++a) {
    for (int b = -1; b <= 1; ++b) {
      table(row, 0) = static_cast<double>(a);
      table(row, 1) = static_cast<double>(b);
      ++row;
    }
  }
  return table;
}

double clamp(double v, double lo, double hi)
{
  return std::min(std::max(v, lo), hi);
}
}  // namespace

const Eigen::Matrix<double, 9, 2> JOINT_ACTIONS = build_joint_actions();

PlanarArm::PlanarArm(const ArmConfig & config)
: config_(config)
{
}

Eigen::Vector2d PlanarArm::forward_kinematics(double q1, double q2) const
{
  const ArmConfig & c = config_;
  return Eigen::Vector2d(
    c.l1 * std::cos(q1) + c.l2 * std::cos(q1 + q2),
    c.l1 * std::sin(q1) + c.l2 * std::sin(q1 + q2));
}

bool PlanarArm::reachable(double x, double y) const
{
  const ArmConfig & c = config_;
  const double r = std::hypot(x, y);
  return std::abs(c.l1 - c.l2) + 1e-6 < r && r < (c.l1 + c.l2) - 1e-6;
}

std::optional<Eigen::Vector2d> PlanarArm::inverse_kinematics(
  double x, double y, bool elbow_up) const
{
  const ArmConfig & c = config_;
  const double r2 = x * x + y * y;
  const double cos_q2 = (r2 - c.l1 * c.l1 - c.l2 * c.l2) / (2.0 * c.l1 * c.l2);
  if (!(cos_q2 >= -1.0 && cos_q2 <= 1.0)) {return std::nullopt;}

  double q2 = std::acos(cos_q2);
  if (!elbow_up) {q2 = -q2;}
  const double q1 = std::atan2(y, x) -
    std::atan2(c.l2 * std::sin(q2), c.l1 + c.l2 * std::cos(q2));

  if (q1 < c.q1_min || q1 > c.q1_max) {return std::nullopt;}
  if (q2 < c.q2_min || q2 > c.q2_max) {return std::nullopt;}
  return Eigen::Vector2d(q1, q2);
}

Eigen::VectorXd PlanarArm::step(const Eigen::VectorXd & state, int action) const
{
  const ArmConfig & c = config_;
  double q1 = state[0], q2 = state[1], px = state[2], py = state[3];

  q1 = clamp(q1 + JOINT_ACTIONS(action, 0) * c.dq, c.q1_min, c.q1_max);
  q2 = clamp(q2 + JOINT_ACTIONS(action, 1) * c.dq, c.q2_min, c.q2_max);

  const Eigen::Vector2d e = forward_kinematics(q1, q2);

  // Push the puck if the gripper is inside contact range.
  if (c.puck_mass_factor > 0.0) {
    double dx = px - e[0], dy = py - e[1];
    double dist = std::hypot(dx, dy);
    if (dist < c.puck_radius) {
      if (dist < 1e-9) {dx = 1.0; dy = 0.0; dist = 1.0;}
      const double overlap = c.puck_radius - dist;
      const double scale = c.puck_mass_factor * overlap / dist;
      px = clamp(px + dx * scale, c.x_min, c.x_max);
      py = clamp(py + dy * scale, c.y_min, c.y_max);
    }
  }

  Eigen::VectorXd out(4);
  out << q1, q2, px, py;
  return out;
}

Eigen::VectorXd PlanarArm::rollout(
  const Eigen::VectorXd & state, const std::vector<int> & actions) const
{
  Eigen::VectorXd s = state;
  for (int a : actions) {s = step(s, a);}
  const Eigen::Vector2d e = forward_kinematics(s[0], s[1]);
  Eigen::VectorXd outcome(4);
  outcome << e[0], e[1], s[2], s[3];
  return outcome;
}

PlanarArm PlanarArm::with_fixed_puck() const
{
  ArmConfig c = config_;
  c.puck_mass_factor = 0.0;
  return PlanarArm(c);
}

}  // namespace intrinsic_core

File: intrinsic_core/intrinsic_core/arm2d.py
"""
A 2-link planar arm on a table with a puck.

This is deliberately the smallest system in which the paper's box
experiment (section 4.5.4) still makes sense.  The four conditions there
were: the box is pushable or not, and perceivable or not.  All four are
reproducible here by changing `puck_mass_factor` and by dropping the puck
dimensions from the discretiser.

State vector:      [q1, q2, px, py]     joint angles, puck position
Outcome vector:    [ex, ey, px, py]     end-effector position, puck position

The outcome deliberately includes the puck.  That is the entire mechanism:
near the puck, two different action sequences that put the gripper in the
same place can leave the puck in different places, so they count as
different outcomes.  Away from the puck they collapse into one.  Nobody
tells the arm the puck is interesting; it falls out of the counting.
"""

from __future__ import annotations

from dataclasses import dataclass, replace
from typing import Sequence

import numpy as np

__all__ = ["ArmConfig", "PlanarArm", "JOINT_ACTIONS"]


# Nine actions: each joint moves down / holds / moves up.
JOINT_ACTIONS = np.array(
    [(a, b) for a in (-1, 0, 1) for b in (-1, 0, 1)],
    dtype=np.float64,
)


@dataclass
class ArmConfig:
    l1: float = 0.30
    l2: float = 0.25

    dq: float = 0.12
    """Radians per step per joint.  The paper's 'power constraint' in
    section 4.7.4 -- and it matters as much here as it does there."""

    q1_limits: tuple[float, float] = (-2.6, 2.6)
    q2_limits: tuple[float, float] = (-2.4, 2.4)

    puck_radius: float = 0.045
    """Contact distance.  Inside this, the gripper pushes the puck."""

    puck_mass_factor: float = 1.0
    """1.0 = puck moves fully with the gripper.
    0.0 = puck is bolted down (the paper's immovable-box condition)."""

    table_bounds: tuple[float, float, float, float] = (-0.55, 0.55, -0.10, 0.55)
    """(xmin, xmax, ymin, ymax) that the puck cannot leave."""


class PlanarArm:
    """Deterministic forward model.  No physics engine, no ROS, no learning."""

    def __init__(self, config: ArmConfig | None = None) -> None:
        self.config = config or ArmConfig()

    # -- kinematics ---------------------------------------------------------

    def forward_kinematics(self, q1: float, q2: float) -> np.ndarray:
        c = self.config
        x = c.l1 * np.cos(q1) + c.l2 * np.cos(q1 + q2)
        y = c.l1 * np.sin(q1) + c.l2 * np.sin(q1 + q2)
        return np.array([x, y])

    def reachable(self, x: float, y: float) -> bool:
        c = self.config
        r = np.hypot(x, y)
        return abs(c.l1 - c.l2) + 1e-6 < r < (c.l1 + c.l2) - 1e-6

    def inverse_kinematics(self, x: float, y: float, elbow_up: bool = True):
        """Returns (q1, q2) or None if unreachable."""
        c = self.config
        r2 = x * x + y * y
        cos_q2 = (r2 - c.l1 ** 2 - c.l2 ** 2) / (2 * c.l1 * c.l2)
        if not -1.0 <= cos_q2 <= 1.0:
            return None
        q2 = np.arccos(cos_q2)
        if not elbow_up:
            q2 = -q2
        q1 = np.arctan2(y, x) - np.arctan2(
            c.l2 * np.sin(q2), c.l1 + c.l2 * np.cos(q2)
        )
        if not (c.q1_limits[0] <= q1 <= c.q1_limits[1]):
            return None
        if not (c.q2_limits[0] <= q2 <= c.q2_limits[1]):
            return None
        return float(q1), float(q2)

    # -- dynamics -----------------------------------------------------------

    def step(self, state: np.ndarray, action: int) -> np.ndarray:
        """One control step.  state = [q1, q2, px, py]."""
        c = self.config
        q1, q2, px, py = (float(v) for v in state)

        dq1, dq2 = JOINT_ACTIONS[int(action)] * c.dq
        q1 = float(np.clip(q1 + dq1, *c.q1_limits))
        q2 = float(np.clip(q2 + dq2, *c.q2_limits))

        ex, ey = self.forward_kinematics(q1, q2)

        # Push the puck if the gripper is inside contact range.
        if c.puck_mass_factor > 0.0:
            dx, dy = px - ex, py - ey
            dist = float(np.hypot(dx, dy))
            if dist < c.puck_radius:
                if dist < 1e-9:
                    dx, dy, dist = 1.0, 0.0, 1.0
                overlap = c.puck_radius - dist
                scale = c.puck_mass_factor * overlap / dist
                px += dx * scale
                py += dy * scale
                xmin, xmax, ymin, ymax = c.table_bounds
                px = float(np.clip(px, xmin, xmax))
                py = float(np.clip(py, ymin, ymax))

        return np.array([q1, q2, px, py])

    def rollout(self, state: np.ndarray, actions: Sequence[int]) -> np.ndarray:
        """Run a sequence of actions and return the OUTCOME vector."""
        s = np.asarray(state, dtype=np.float64).copy()
        for a in actions:
            s = self.step(s, int(a))
        ex, ey = self.forward_kinematics(s[0], s[1])
        return np.array([ex, ey, s[2], s[3]])

    # -- variants of the paper's box experiment -----------------------------

    def with_fixed_puck(self) -> "PlanarArm":
        """The immovable-box condition: present, but cannot be pushed."""
        return PlanarArm(replace(self.config, puck_mass_factor=0.0))
